### PR TITLE
Implementing list diffing

### PR DESCRIPTION
**Pull request by samabbey-sonocent**
Implementing a more sophisticated list diff using the Wagner–Fischer algorithm. Example [here](https://www.google.com/search?q=wagnerfisher+diff&rlz=1C5CHFA_enGB891GB891&oq=wagn&aqs=chrome.0.69i59j46i433j69i57j0j69i59j69i60l3.1319j0j7&sourceid=chrome&ie=UTF-8).

### CODE REVIEW

**Comment by samabbey-sonocent** [comment]
I wondered what you thought about adding some property based tests? i.e test that diffing two randomly generated pieces of Json and applying the diffs again result in the new Json. This project seems to lend itself nicely to that

File: test/json_patch_test.dart [hunk 1]
```diff
@@ -1,10 +1,10 @@
 import 'dart:convert';
 
-import 'package:test/test.dart';
 import 'package:json_patch/json_patch.dart';
+import 'package:test/test.dart';
 
 void main() {
-  group('JsonPatch', () {
+  group('.diff', () {
```
Comment (pikaju) [inline test/json_patch_test.dart]: This `'JsonPatch'` was deliberate, as it is the class being tested, and each test had `'.diff'` to describe which method is being tested for what.

File: lib/src/list_edit_matrix.dart [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import 'dart:math';
+
+import 'package:collection/collection.dart';
+
+/// Utility class for diffing lists using the Wagner–Fischer algorithm.
+/// It creates a matrix of operations which need to be applied to the old list
+/// to recreate the new one.
```
Comment (pikaju) [inline lib/src/list_edit_matrix.dart:9]: I think you put these comments in front of the wrong type.

File: lib/src/json_patch.dart [hunk 3]
```diff
@@ -45,7 +47,7 @@ class JsonPatch {
         ];
       }
 
-      // For primitive types, use the == operator for comparisson and replace if necessary.
+      // For primitive types, use the == operator for comparison and replace if necessary.
```
Comment (pikaju) [inline lib/src/json_patch.dart:50]: Sneaky.

File: test/list_edit_matrix_test.dart [hunk 4]
```diff
@@ -0,0 +1,98 @@
+import 'package:collection/collection.dart';
+import 'package:json_patch/src/list_edit_matrix.dart';
+import 'package:test/test.dart';
+
+main() {
+  group('buildEditMatrix', () {
+    test('buildEditMatrix should handle empty lists', () {
```
Comment (pikaju) [inline test/list_edit_matrix_test.dart]: Would be nice to follow the same naming convention I mentioned earlier, i.e. "ListEditMatrix" group and ".buildEditMatrix handles empty list"

**Review by pikaju — CHANGES_REQUESTED** [review]
Just some minor style complaints, thank you for this! I'll go ahead and trust you that the implementation is correct because I don't understand much without going in-depth.

File: test/json_patch_test.dart (hunk 1 above)
Comment (samabbey-sonocent) [reply test/json_patch_test.dart]: Ah cool! I'll change that back 😄

File: lib/src/list_edit_matrix.dart (hunk 2 above)
Comment (samabbey-sonocent) [reply lib/src/list_edit_matrix.dart:9]: 👍

File: test/list_edit_matrix_test.dart (hunk 4 above)
Comment (samabbey-sonocent) [reply test/list_edit_matrix_test.dart]: 👍

**Comment by samabbey-sonocent** [comment]
Amazing! Thanks for getting round to looking at this so quickly! A couple of my colleagues have had a look at the implementation so it has had a few eyes on it don't worry!

**Comment by pikaju** [comment]
Okay cool, thank you for the quick changes. I'll go ahead and publish a new version.

**Pull request merged**